### PR TITLE
Add tool calling support for ChatModel

### PR DESCRIPTION
**Pull request by daniellok-db**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

This PR adds basic tool calling functionality for ChatModel, matching the [OpenAI spec](https://platform.openai.com/docs/api-reference/chat/create). Here's a screenshot reference for the created classes:

## Messages schema

Screenshot 2024-09-19 at 1 01 20 PM

## Tool Definitions

Top level (the API spec doesn't go completely into details)
Screenshot 2024-09-19 at 4 56 19 PM

Deep dive to params schema

Screenshot 2024-09-19 at 5 13 52 PM

The PR adds the required dataclasses, and updates the chatmodel input/output schemas. Check out the gists below to see how the workflow looks.

More documentation is definitely needed for ChatModel, I'll add a specific section to the docs in a follow up PR.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Tested a basic tool calling workflow with `gpt-4o-mini`. The scripts are a bit long for a PR description so here are some gists:

1. Defining a tool calling ChatModel: 
  a. This is a toy model that simply forwards the user input to OpenAI
  b. https://gist.github.com/daniellok-db/43aca672266bcdc1eae42db470c415e5
2. Interacting with a tool calling ChatModel:
  a. This script demonstrates two rounds of message passing, involving a tool call by the model to get the weather
  b. https://gist.github.com/daniellok-db/3337f78d3f93d0260cccab0b41f940b9

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

`mlflow.pyfunc.ChatModel` now supports tool calling! Check out the documentation for details.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/types/llm.py [hunk 1]
```diff
@@ -13,6 +13,9 @@
 #       is not supported, so the code here is a little ugly.
 
 
+JSON_SCHEMA_TYPES = ["string", "number", "integer", "object", "array", "boolean", "null"]
```
Comment (daniellok-db) [inline mlflow/types/llm.py:16]: https://json-schema.org/understanding-json-schema/reference/type#type-specific-keywords

File: mlflow/types/llm.py [hunk 2]
```diff
@@ -90,24 +127,73 @@ def from_dict(cls, data):
...
+
+    name: str
+    arguments: str
+
+    def __post_init__(self):
+        self._validate_field("name", str, True)
+        self._validate_field("arguments", str, True)
+
+    def to_tool_call(self, id):
+        return ToolCall(id=id, function=self)
```
Comment (daniellok-db) [inline mlflow/types/llm.py:148]: This is a convenience method so that users don't have to manually construct a `ToolCall` class. At the moment, the `ToolCall` class is a necessary but useless wrapper for the actual tool call arguments.

File: mlflow/types/llm.py [hunk 3]
```diff
@@ -90,24 +127,73 @@ def from_dict(cls, data):
...
+
+@dataclass
+class ToolCall(_BaseDataclass):
+    """
+    A tool call made by the model.
+
+    Args:
+        id (str): The ID of the tool call.
+        function (:py:class:`FunctionToolCallArguments`): The arguments of the function tool call.
+        type (str): The type of the object. Currently only "function" is supported.
```
Comment (daniellok-db) [inline mlflow/types/llm.py:159]: As mentioned above, right now this class is just necessary to conform to the OpenAI spec, see the [API spec](https://platform.openai.com/docs/api-reference/chat/object), `choices > message > tool_calls`

Screenshot 2024-09-19 at 1 01 20 PM

In the future, I imagine that they may add other types besides `function`, and then this class will have more utility.

File: mlflow/types/llm.py [hunk 4]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
+        self._validate_field("name", str, True)
+        self._validate_field("description", str, False)
+        self._convert_dataclass("parameters", ToolParamsSchema, False)
+        self._validate_field("strict", bool, False)
+
+    def to_tool_definition(self):
+        """
+        Convenience function for wrapping this in a ToolDefiniton
+        """
+        return ToolDefinition(type="function", function=self)
```
Comment (daniellok-db) [inline mlflow/types/llm.py:307]: Similar to above, this prevents needing to interact with `ToolDefinition` directly since it's just a wrapper

File: mlflow/types/llm.py [hunk 5]
```diff
@@ -395,8 +602,31 @@ def __post_init__(self):
                 Object(
                     [
                         Property("role", DataType.string),
-                        Property("content", DataType.string),
+                        Property("content", DataType.string, False),
```
Comment (daniellok-db) [inline mlflow/types/llm.py:605]: `content` is not required if `refusal` or `tool_calls` is present

File: mlflow/types/llm.py [hunk 6]
```diff
@@ -90,24 +127,73 @@ def from_dict(cls, data):
...
+    """
+
+    name: str
+    arguments: str
+
+    def __post_init__(self):
+        self._validate_field("name", str, True)
+        self._validate_field("arguments", str, True)
+
+    def to_tool_call(self, id):
```
Comment (serena-ruan) [inline mlflow/types/llm.py:147]: Why do we need this?

File: mlflow/types/llm.py [hunk 7]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
+        self._convert_dataclass_map("properties", ParamProperty, True)
+        self._validate_literal("type", ["object"], True)
+        self._validate_list("required", str, False)
+        self._validate_field("additionalProperties", bool, False)
+
+
+@dataclass
+class FunctionToolDefinition(_BaseDataclass):
+    """
+    Definition for function tools (currently the only supported type of tool).
```
Comment (serena-ruan) [inline mlflow/types/llm.py:279]: Could we add links to where those definition refers to? It's quite hard to follow what is needed and added

File: mlflow/types/llm.py [hunk 8]
```diff
@@ -410,6 +640,72 @@ def __post_init__(self):
         ColSpec(name="top_k", type=DataType.long, required=False),
         ColSpec(name="frequency_penalty", type=DataType.double, required=False),
         ColSpec(name="presence_penalty", type=DataType.double, required=False),
+        ColSpec(
```
Comment (harupy) [inline mlflow/types/llm.py:643]: Feel free to turn off formatting if that makes it easier to read this block.

File: mlflow/types/llm.py [hunk 9]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
+@dataclass
+class FunctionToolDefinition(_BaseDataclass):
+    """
+    Definition for function tools (currently the only supported type of tool).
+
+    Args:
+        name (str): The name of the tool.
+        description (str): A description of what the tool does, and how it should be used.
+            **Optional**, defaults to ``None``
+        parameters (Dict[str, :py:class:`ToolParamsSchema`]): A mapping of parameter names to their
```
Comment (harupy) [inline mlflow/types/llm.py:285]:
```suggestion
        parameters: A mapping of parameter names to their
```

I'd remove the type because there is a type hint.

File: mlflow/types/llm.py [hunk 10]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
+            definitions. If not provided, this defines a function without parameters.
+            **Optional**, defaults to ``None``
+        strict (bool): Whether or not to opt into structured outputs.
+            **Optional**, defaults to ``None``.
+    """
+
+    name: str
+    description: Optional[str] = None
+    parameters: ToolParamsSchema = None
```
Comment (harupy) [inline mlflow/types/llm.py:294]:
```suggestion
    parameters: Optional[ToolParamsSchema] = None
```

File: mlflow/types/llm.py [hunk 11]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
+    strict: Optional[bool] = None
```
Comment (harupy) [inline mlflow/types/llm.py:295]: what does `None` mean?

File: mlflow/types/llm.py [hunk 12]
```diff
@@ -90,24 +127,73 @@ def from_dict(cls, data):
...
     """
     A message in a chat request or response.
 
     Args:
-        role (str): The role of the entity that sent the message (e.g. ``"user"``, ``"system"``).
+        role (str): The role of the entity that sent the message (e.g. ``"user"``,
+            ``"system"``, ``"assistant"``, ``"tool"``).
         content (str): The content of the message.
-            **Optional** Supplied if a non-refusal response is provided.
+            **Optional** Can be null if refusal or tool_calls are provided.
```
Comment (harupy) [inline mlflow/types/llm.py:181]:
```suggestion
            **Optional** Can be None if refusal or tool_calls are provided.
```

File: mlflow/types/llm.py [hunk 13]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
             self._validate_field("content", str, True)
 
         self._validate_field("name", str, False)
+        self._convert_dataclass_list("tool_calls", ToolCall, False)
+        self._validate_field("tool_call_id", str, False)
+
+
+@dataclass
+class ParamType(_BaseDataclass):
+    type: Literal["string", "number", "integer", "object", "array", "boolean", "null"]
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/types/llm.py:217]: Is `Literal[tuple(JSON_SCHEMA_TYPES)]` really valid?

File: mlflow/types/llm.py [hunk 14]
```diff
@@ -116,10 +202,127 @@ def __post_init__(self):
...
+            **Optional**, defaults to ``None``
+        enum (List[str]): Used to constrain the possible values for the parameter.
+            **Optional**, defaults to ``None``
+        items (:py:class:`ParamProperty`): If the param is of ``array`` type, this field can be
+            used to specify the type of its items. **Optional**, defaults to ``None``
+    """
+
+    description: Optional[str] = None
+    enum: Optional[List[str]] = None
+    items: Optional[ParamType] = None
```
Comment (harupy) [inline mlflow/types/llm.py:241]: Is `type` missing?

File: mlflow/types/llm.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (daniellok-db) [reply mlflow/types/llm.py:217]: it seems to work but it feels wrong
```

File: mlflow/types/llm.py (hunk 6 above)
Comment (daniellok-db) [reply mlflow/types/llm.py:147]: I thought it was easier than users calling

```python
# without convenience method
tool_call_args = FunctionToolCallArguments(name="get_weather", arguments="")
tool_call = ToolCall(id=id, function=tool_call_args)

# with
tool_call = FunctionToolCallArguments(name="get_weather", arguments="").to_tool_call(id)
```

Eventually as OpenAI (or maybe Databricks or some other provider) adds more types of tools, this `to_tool_call()` method may be more helpful than requiring the user to remember what args `ToolCall` needs to be provided with

```python
# without the method, users need to specify `uc_function`, not `function`
tool_call_args = UnityCatalogToolCallArguments(name="some_tool", catalog="main.ml.tools")
tool_call = ToolCall(id=id, uc_function=tool_call_args)

# with method, they don't need to think
UnityCatalogToolCallArguments(name="some_tool", catalog="main.ml.tools").to_tool_call()
```

File: mlflow/types/llm.py (hunk 7 above)
Comment (daniellok-db) [reply mlflow/types/llm.py:279]: Agree! Let me add this to the PR description. I'm not sure if i should add it as a docstring because although we're modelling after OpenAI it might not be guaranteed that we will always follow that standard (e.g. we might expand the signature to be compatible with more providers, or allow users to override)

File: mlflow/types/llm.py (hunk 5 above)
Comment (serena-ruan) [reply mlflow/types/llm.py:607]: This sounds like another tricky case we need to consider in model signature :)

File: mlflow/types/llm.py [hunk 15]
```diff
@@ -90,24 +127,73 @@ def from_dict(cls, data):
...
+    """
+    A tool call made by the model.
+
+    Args:
+        id (str): The ID of the tool call.
+        function (:py:class:`FunctionToolCallArguments`): The arguments of the function tool call.
+        type (str): The type of the object. Currently only "function" is supported.
+    """
+
+    id: str
```
Comment (serena-ruan) [inline mlflow/types/llm.py:162]: Wondering if we can generate a default value here? Otherwise `to_tool_call` still need to pass the id 🤔

**Review by serena-ruan — APPROVED** [review]
Overall LGTM! Left some nits

File: mlflow/pyfunc/loaders/chat_model.py [hunk 16]
```diff
@@ -45,7 +46,8 @@ def _convert_input(self, model_input):
             dict_input = model_input
         elif isinstance(model_input, pandas.DataFrame):
             dict_input = {
-                key: value[0] for key, value in model_input.to_dict(orient="list").items()
+                k: _convert_llm_ndarray_to_list(v[0]) if isinstance(v[0], list) else v[0]
```
[the remark this answers is not in the document]
Comment (daniellok-db) [reply mlflow/pyfunc/loaders/chat_model.py:49]: ahh got it, thank you!

File: mlflow/types/llm.py (hunk 15 above)
Comment (daniellok-db) [reply mlflow/types/llm.py:162]: yeah i think we can do that—it seems the actual ID is not so important here, the client just needs a way to reference the tool call message. i'll add it!

**Review by bbqiu — APPROVED** [review]
this LGTM from a tool calling perspective -- the new ChatMessage will be able to contain all the information from LangChain HumanMessage/AIMessage/ToolMessage types
